src/layers/domain/core/questionnaire.py

```diff
@@ -1,75 +1,149 @@
+            raise ValueError
+        return type
 
 
 class Questionnaire(BaseModel):
     """
     A Questionnaire represents a collection of Questions, in a specific order.
     """
 
-    id: UUID
     name: str = Field(regex=ENTITY_NAME_REGEX)
-    _questions: list[Question] = (lambda: [])()
+    version: int
+    questions: dict[str, Question] = Field(default_factory=dict)
 
     def __contains__(self, question_name: str) -> bool:
         """
         Returns true if the question specified exists within the questionnaire
         """
-        return question_name in (q.name for q in self._questions)
+
+        return question_name in self.questions
 
     def add_question(
         self,
         name: str,
-        type: QuestionType = QuestionType.STRING,
+        type: Type = str,
         multiple: bool = False,
+        validation_rules: set[FunctionType] = None,
+        choices: set[T] = None,
     ):
         """
-        Adds a new question to the questionnaire.  Once a questionnaire is
-        locked this method will throw an error.
+        Adds a new question to the questionnaire.
         """
-        if name in self:
-            raise DuplicateError(f"Question exists: {name}")
-        result = Question(name=name, type=type, multiple=multiple)
-        self._questions.append(result)
-        return result
+        if name in self.questions:
+            raise DuplicateError(f"Question '{name}' already exists.")
 
-    def remove_question(self, name: str):
-        """
-        Removes a question from the questionnaire.  Once a questionnaire is
-        locked this method will throw an error.
-        """
-        raise NotImplementedError()
+        # Validate choices are of the same type as the question type
+        if choices is not None and not all(
+            isinstance(choice, type) for choice in choices
+        ):
+            raise ValueError(
+                f"Choices must be of the same type as the question type: {type}"
+            )
 
-    def revise(self) -> "Questionnaire":
-        """
-        Creates an exact duplicate of the questionnaire, which is the only way
-        to edit questionnaires once they become locked.
-        """
-        raise NotImplementedError()
+        question = Question[type](
+            name=name,
+            type=type,
+            multiple=multiple,
+            validation_rules=validation_rules,
+            choices=choices,
+        )
+
+        self.questions[name] = question
+        return question
+
+
+class QuestionnaireResponse(BaseModel):
+    """
+    Validates questionnaire responses against questionnaire questions
+    """
+
+    questionnaire: Questionnaire
+    responses: list[tuple[str, list]]
+
+    @validator("responses", each_item=True)
+    def validate_responses(
+        response: tuple[str, list], values: dict[str, Questionnaire]
+    ):
+        questionnaire = values.get("questionnaire")
+        question_name = response[0]
```

Comment:
rather than passing the response tuple into both functions better to unpack here

```python
question_name, answers = response
```

and then down below do

```
validate_response_against_question(question=question, answers=answers)
```